coderd/healthcheck/websocket.go

```diff
@@ -2,11 +2,120 @@ package healthcheck
 
 import (
 	"context"
+	"io"
+	"net/http"
 	"net/url"
+	"strconv"
+	"time"
+
+	"golang.org/x/xerrors"
+	"nhooyr.io/websocket"
+
+	"github.com/coder/coder/coderd/httpapi"
 )
 
-type WebsocketReport struct{}
+type WebsocketReportOptions struct {
+	APIKey     string
+	AccessURL  *url.URL
+	HTTPClient *http.Client
+}
+
+type WebsocketReport struct {
+	Error error `json:"error"`
+}
+
+func (r *WebsocketReport) Run(ctx context.Context, opts *WebsocketReportOptions) {
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	u, err := opts.AccessURL.Parse("/api/v2/debug/ws")
+	if err != nil {
+		r.Error = xerrors.Errorf("parse access url: %w", err)
+		return
+	}
+	if u.Scheme == "https" {
+		u.Scheme = "wss"
+	} else {
+		u.Scheme = "ws"
+	}
+
+	//nolint:bodyclose
+	c, _, err := websocket.Dial(ctx, u.String(), &websocket.DialOptions{
+		HTTPClient: opts.HTTPClient,
+		HTTPHeader: http.Header{"Coder-Session-Token": []string{opts.APIKey}},
+	})
+	if err != nil {
+		r.Error = xerrors.Errorf("websocket dial: %w", err)
+		return
+	}
+	defer c.Close(websocket.StatusGoingAway, "goodbye")
+
+	for i := 0; i < 3; i++ {
+		msg := strconv.Itoa(i)
+		err := c.Write(ctx, websocket.MessageText, []byte(msg))
+		if err != nil {
+			r.Error = xerrors.Errorf("write message: %w", err)
+			return
+		}
+
+		ty, got, err := c.Read(ctx)
+		if err != nil {
+			r.Error = xerrors.Errorf("read message: %w", err)
+			return
+		}
+
+		if ty != websocket.MessageText {
+			r.Error = xerrors.Errorf("received incorrect message type: %v", ty)
+			return
+		}
+
+		if string(got) != msg {
+			r.Error = xerrors.Errorf("received incorrect message: wanted %q, got %q", msg, string(got))
+			return
+		}
+	}
+
+	c.Close(websocket.StatusGoingAway, "goodbye")
+}
+
+type WebsocketEchoServer struct{}
+
+func (WebsocketEchoServer) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	c, err := websocket.Accept(rw, r, &websocket.AcceptOptions{})
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, "Invalid websocket.")
```

Comment:
Should reply with the error from above because it may be useful for debugging e.g. "the host doesn't match origin". And then make the client code read the error